File: app/src/main/java/com/bove/martin/manossolidarias/activities/fragments/OngInfoFragment.java
package com.bove.martin.manossolidarias.activities.fragments;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.view.LayoutInflaterCompat;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.bove.martin.manossolidarias.R;
import com.bove.martin.manossolidarias.activities.base.BaseActivity;
import com.bove.martin.manossolidarias.activities.interfaces.FragmentComunication;
import com.bove.martin.manossolidarias.activities.utils.CircleTransform;
import com.bove.martin.manossolidarias.model.Institucion;
import com.google.gson.Gson;
import com.mikepenz.iconics.context.IconicsContextWrapper;
import com.mikepenz.iconics.context.IconicsLayoutInflater;
import com.mikepenz.iconics.context.IconicsLayoutInflater2;
import com.mikepenz.iconics.view.IconicsImageView;
import com.squareup.picasso.Picasso;

import org.w3c.dom.Text;

public class OngInfoFragment extends Fragment {
    private FragmentComunication callback;
    private Institucion ong;

    private ImageView header;
    private ImageView logo;
    private TextView mision;
    private TextView horarioTitulo;
    private TextView horario;
    private TextView miscTitulo;
    private TextView misc;
    private IconicsImageView teleIcon;
    private TextView tele;
    private View dividerTele;
    private IconicsImageView mailIcon;
    private TextView mail;
    private View dividerMail;
    private IconicsImageView webIcon;
    private TextView web;
    private View dividerWeb;
    private IconicsImageView facebookIcon;
    private IconicsImageView twitterIcon;
    private ImageView instagramIcon;
    private IconicsImageView youtubeIcon;

    public SharedPreferences preferences;

    public OngInfoFragment() {
        // Required empty public constructor
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        try {
            callback = (FragmentComunication) context;
        } catch (Exception e) {
            throw new ClassCastException(context.toString() + " Debes implementar Fragament Comunication.");
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        final View view =  inflater.inflate(R.layout.fragment_ong_info, container, false);

        getCurrentONG();

        // Instaciamos los elementos
        header = view.findViewById(R.id.infoOngHeader);
        logo = view.findViewById(R.id.infoOngLogo);
        mision = view.findViewById(R.id.infoMision);
        horarioTitulo = view.findViewById(R.id.infoHorarioTittle);
        horario = view.findViewById(R.id.infoHorario);
        miscTitulo = view.findViewById(R.id.infoMiscTittle);
        misc = view.findViewById(R.id.infoMisc);

        teleIcon = view.findViewById(R.id.OngTeleiIcon);
        tele = view.findViewById(R.id.ongInfoTel);
        dividerTele = view.findViewById(R.id.divider1);

        mailIcon = view.findViewById(R.id.OngMailiIcon);
        mail = view.findViewById(R.id.ongInfoEmail);
        dividerMail = view.findViewById(R.id.divider2);

        webIcon = view.findViewById(R.id.OngWebiIcon);
        web = view.findViewById(R.id.ongInfoWeb);
        dividerWeb = view.findViewById(R.id.divider3);

        facebookIcon = view.findViewById(R.id.iconicsImageFB);
        instagramIcon = view.findViewById(R.id.iconicsImageInstagram);
        twitterIcon = view.findViewById(R.id.iconicsImageTwitter);
        youtubeIcon = view.findViewById(R.id.iconicsImageYoutube);

        // Cargamos los elementos
        // TODO hacer un placeholder de manos solidarias para el header o implementar un spinner
        Picasso.with(view.getContext()).load(ong.getHeader_img_url()).placeholder(R.drawable.placeholder).fit().centerCrop().into(header);
        Picasso.with(view.getContext()).load(ong.getLogo_url()).transform(new CircleTransform()).fit().into(logo);
        mision.setText(ong.getDescripcion());

        if(TextUtils.isEmpty(ong.getHorario())) {
            horarioTitulo.setVisibility(View.GONE);
            horario.setVisibility(View.GONE);
        } else {
            horario.setText(ong.getHorario());
        }

        if(TextUtils.isEmpty(ong.getMisc())) {
            miscTitulo.setVisibility(View.GONE);
            misc.setVisibility(View.GONE);
        } else {
            misc.setText(ong.getMisc());
        }

        if(TextUtils.isEmpty(ong.getTelefono())) {
            teleIcon.setVisibility(View.GONE);
            tele.setVisibility(View.GONE);
            dividerTele.setVisibility(View.GONE);
        } else {
            tele.setText(ong.getTelefono());
            tele.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    makeCall(ong.getTelefono());
                }
            });
        }

        if(TextUtils.isEmpty(ong.getEmail())) {
            mailIcon.setVisibility(View.GONE);
            mail.setVisibility(View.GONE);
            dividerMail.setVisibility(View.GONE);
        } else {
            mail.setText(ong.getEmail());
            mail.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    openMail(ong.getEmail());
                }
            });
        }

        if(TextUtils.isEmpty(ong.getWeb())) {
            webIcon.setVisibility(View.GONE);
            web.setVisibility(View.GONE);
        } else {
            web.setText(ong.getWeb());
            web.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    openWeb(ong.getWeb());
                }
            });
        }

        // Social section
        if(TextUtils.isEmpty(ong.getFacebook())) {
           facebookIcon.setVisibility(View.GONE);
        } else {
            facebookIcon.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    openLink(ong.getFacebook(), "facebook");
                }
            });
        }
        if(TextUtils.isEmpty(ong.getInstagram())) {
            instagramIcon.setVisibility(View.GONE);
        } else  {
            instagramIcon.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    openLink(ong.getInstagram(), "instagram");
                }
            });
        }
        if(TextUtils.isEmpty(ong.getTwitter())) {
            twitterIcon.setVisibility(View.GONE);
        } else {
            twitterIcon.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    openLink(ong.getTwitter(), "twitter");
                }
            });
        }
        if(TextUtils.isEmpty(ong.getYoutube())) {
            youtubeIcon.setVisibility(View.GONE);
        } else {
            youtubeIcon.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    openLink(ong.getYoutube(), "youtube");
                }
            });
        }

        return view;
    }

    private void getCurrentONG() {
        if (ong == null ) {
            if (preferences == null) {
                // Obtenemos el objeto guardado previamente en las pref
                preferences = this.getActivity().getSharedPreferences(BaseActivity.SHARED_PREF, Context.MODE_PRIVATE);
            }
            Gson gson = new Gson();
            String json = preferences.getString("institucion", "");
            ong = gson.fromJson(json, Institucion.class);
        }
    }

    private void makeCall(String telefono) {
        Intent intent = new Intent(Intent.ACTION_DIAL);
        intent.setData(Uri.parse("tel:" + telefono));
        startActivity(intent);
    }

    private void openMail(String emial) {
        Intent emailIntent = new Intent(Intent.ACTION_SENDTO);
        emailIntent.setData(Uri.parse("mailto:" + emial));
        startActivity(emailIntent);
    }

    private void openWeb(String web) {
        Intent i = new Intent(Intent.ACTION_VIEW);
        i.setData(Uri.parse(web));
        try {
            startActivity(i);
        }catch (ActivityNotFoundException e) {
            Toast.makeText(getActivity(), getString(R.string.error_no_intent), Toast.LENGTH_SHORT).show();
        }
    }

    // Open links in native apps
    private void openLink(String link, String service) {
        String base_url = "";
        String service_url = "";
        String paquete = "";

        // TODO revisar que se habra bien en facebook
        switch (service) {
            case "facebook": {
                base_url =  "https://m.facebook.com/";
                paquete = "";
                service_url = "";
                break;
            }
            case "twitter": {
                base_url =  "https://twitter.com/";
                paquete = "com.twitter.android";
                service_url = "http://twitter.com/";
                break;
            }
            case "instagram": {
                base_url =  "http://instagram.com/";
                paquete = "com.instagram.android";
                service_url = "http://instagram.com/_u/";
                break;
            }
            case "youtube": {
                base_url =  "https://www.youtube.com/user/";
                paquete = "com.google.android.youtube";
                service_url = "https://www.youtube.com/user/";
                break;
            }
        }

        Uri uri = Uri.parse(service_url + link);
        Intent likeIng = new Intent(Intent.ACTION_VIEW, uri);
        likeIng.setPackage(paquete);

        try {
            startActivity(likeIng);
        } catch (ActivityNotFoundException e) {
            openWeb(base_url + link);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        getCurrentONG();
    }

    @Override
    public void onPause() {
        super.onPause();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
    }
}

File: app/src/main/java/com/bove/martin/manossolidarias/presentation/fragments/OngInfoFragment.kt
package com.bove.martin.manossolidarias.presentation.fragments

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import com.bove.martin.manossolidarias.R

import com.bove.martin.manossolidarias.databinding.FragmentOngInfoBinding
import com.bove.martin.manossolidarias.domain.model.Institucion
import com.bove.martin.manossolidarias.presentation.OngInfoActivity

import com.bumptech.glide.Glide
import com.google.gson.Gson

class OngInfoFragment : Fragment() {
    private lateinit var ong: Institucion
    private lateinit var preferences: SharedPreferences

    private var _binding: FragmentOngInfoBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        _binding = FragmentOngInfoBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        preferences = (activity as OngInfoActivity).getSharedPreferences()!!
        loadOng()

        // Cargamos los elementos
        Glide.with(this)
                .load(ong.header_img_url)
                .centerCrop()
                .placeholder(R.drawable.placeholder)
                .into(binding.infoOngHeader)

        Glide.with(this)
                .load(ong.logo_url)
                .circleCrop()
                .placeholder(R.drawable.oval_place_holder)
                .into(binding.infoOngLogo)

        binding.infoMision.text = ong.descripcion
        if (TextUtils.isEmpty(ong.horario)) {
            binding.infoHorarioTittle.visibility = View.GONE
            binding.infoHorario.visibility = View.GONE
        } else {
            binding.infoHorario.text = ong.horario
        }
        if (TextUtils.isEmpty(ong.direccion)) {
            binding.infoDireccionTittle.visibility = View.GONE
            binding.infoDireccion.visibility = View.GONE
        } else {
            binding.infoDireccion.text = ong.direccion
            //fixme solucionar la comunicacion con el activity usando viewmodels
           // binding.infoDireccion.setOnClickListener { callback!!.changeTab(OngInfoActivity.MAP_FRAGMENT) }
        }
        if (TextUtils.isEmpty(ong.misc)) {
            binding.infoMiscTittle.visibility = View.GONE
            binding.infoMisc.visibility = View.GONE
        } else {
            binding.infoMisc.text = ong.misc
        }
        if (TextUtils.isEmpty(ong.telefono)) {
            binding.OngTeleiIcon.visibility = View.GONE
            binding.ongInfoTel.visibility = View.GONE
            binding.divider1.visibility = View.GONE
        } else {
            binding.ongInfoTel.text = ong.telefono
            binding.ongInfoTel.setOnClickListener { makeCall(ong.telefono) }
        }
        if (TextUtils.isEmpty(ong.email)) {
            binding.OngMailiIcon.visibility = View.GONE
            binding.ongInfoEmail.visibility = View.GONE
            binding.divider2.visibility = View.GONE
        } else {
            binding.ongInfoEmail.text = ong.email
            binding.ongInfoEmail.setOnClickListener { openMail(ong.email) }
        }
        if (TextUtils.isEmpty(ong.web)) {
            binding.OngWebiIcon.visibility = View.GONE
            binding.ongInfoWeb.visibility = View.GONE
            binding.divider3.visibility = View.GONE
        } else {
            binding.ongInfoWeb.text = ong.web
            binding.ongInfoWeb.setOnClickListener { openWeb(ong.web) }
        }

        // Social section
        if (TextUtils.isEmpty(ong.facebook)) {
            binding.iconicsImageFB.visibility = View.GONE
        } else {
            binding.iconicsImageFB.setOnClickListener { openLink(ong.facebook, "facebook") }
        }
        if (TextUtils.isEmpty(ong.instagram)) {
            binding.iconicsImageInstagram.visibility = View.GONE
        } else {
            binding.iconicsImageInstagram.setOnClickListener { openLink(ong.instagram, "instagram") }
        }
        if (TextUtils.isEmpty(ong.twitter)) {
            binding.iconicsImageTwitter.visibility = View.GONE
        } else {
            binding.iconicsImageTwitter.setOnClickListener { openLink(ong.twitter, "twitter") }
        }
        if (TextUtils.isEmpty(ong.youtube)) {
            binding.iconicsImageYoutube.visibility = View.GONE
        } else {
            binding.iconicsImageYoutube.setOnClickListener { openLink(ong.youtube, "youtube") }
        }
        if (TextUtils.isEmpty(ong.whatsapp)) {
            binding.iconicsImageWhatsapp.visibility = View.GONE
        } else {
            binding.iconicsImageWhatsapp.setOnClickListener { openLink(ong.whatsapp, "whastapp") }
        }
    }

    // Obtenemos el objeto guardado previamente en las pref
    private fun loadOng() {
        val gson = Gson()
        val json = preferences.getString("institucion", "")
        ong =  gson.fromJson(json, Institucion::class.java)
    }

    private fun makeCall(telefono: String?) {
        val intent = Intent(Intent.ACTION_DIAL)
        intent.data = Uri.parse("tel:$telefono")
        startActivity(intent)
    }

    private fun openMail(emial: String?) {
        val emailIntent = Intent(Intent.ACTION_SENDTO)
        emailIntent.data = Uri.parse("mailto:$emial")
        startActivity(emailIntent)
    }

    private fun openWeb(web: String?) {
        val i = Intent(Intent.ACTION_VIEW)
        i.data = Uri.parse(web)
        try {
            startActivity(i)
        } catch (e: ActivityNotFoundException) {
            Toast.makeText(activity, getString(R.string.error_no_intent), Toast.LENGTH_SHORT).show()
        }
    }

    // Open links in native apps
    private fun openLink(link: String?, service: String) {
        var base_url = ""
        var service_url = ""
        var paquete = ""
        var customIntent = false
        when (service) {
            "facebook" -> {
                base_url = "https://m.facebook.com/"
                paquete = ""
                service_url = ""
            }
            "twitter" -> {
                base_url = "https://twitter.com/"
                paquete = "com.twitter.android"
                service_url = "http://twitter.com/"
            }
            "instagram" -> {
                base_url = "http://instagram.com/"
                paquete = "com.instagram.android"
                service_url = "http://instagram.com/_u/"
            }
            "youtube" -> {
                base_url = "https://www.youtube.com/user/"
                paquete = "com.google.android.youtube"
                service_url = "https://www.youtube.com/user/"
            }
            "whastapp" -> {
                customIntent = true
                try {
                    val text = "Hola " + ong.nombre
                    // Formato 543416709663
                    val toNumber = ong.whatsapp
                    val intent = Intent(Intent.ACTION_VIEW)
                    intent.data = Uri.parse("http://api.whatsapp.com/send?phone=$toNumber&text=$text")
                    startActivity(intent)
                } catch (e: Exception) {
                    e.printStackTrace()
                    //Toast.makeText(getContext(), e.getLocalizedMessage(), Toast.LENGTH_LONG).show();
                }
            }
        }
        if (!customIntent) {
            val uri = Uri.parse(service_url + link)
            val likeIng = Intent(Intent.ACTION_VIEW, uri)
            likeIng.setPackage(paquete)
            try {
                startActivity(likeIng)
            } catch (e: ActivityNotFoundException) {
                openWeb(base_url + link)
            }
        }
    }

    // Si se resume debe traer nuevamente el la ong del sharedPref.
    override fun onResume() {
        super.onResume()
        loadOng()
    }

    override fun onPause() {
        super.onPause()
    }

    override fun onDestroy() {
        super.onDestroy()
    }

    override fun onLowMemory() {
        super.onLowMemory()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}